File: Práctica4/lib/BetaPowerEfficientSpaceStation.rb
#encoding:utf-8

require_relative 'Dice'
require_relative 'PowerEfficientSpaceStation'
require_relative 'BetaPowerEfficientSpaceStationToUI'

module Deepspace
  # CLase de estaciones eficientes con tecnología beta
  # Da una mayor potencia de disparo
  class BetaPowerEfficientSpaceStation < PowerEfficientSpaceStation
    # @!attribute [Float] eficiencia extra de la tecnología beta
    @@EXTRAEFFICIENCY = 1.20

    #Constructor
    # @param  station [SpaceStation] estación espacial a convertir en estación eficiente beta
    # --Overriden
    def initialize(station)
      super(station)

      # @!attribute [Dice] dado
      @dice = Dice.new
    end

    # Realiza un disparo
    # @return [Float] potencia de disparo
    def fire
      if @dice.extraEfficiency
        @@EXTRAEFFICIENCY*super
      else
        super
      end
    end

    # String representation
    # @return [String] string representation
    def to_s
      message = "(Beta) " + super
      return message
    end

    # To UI
    def getUIversion
      return BetaPowerEfficientSpaceStationToUI.new(self)
    end
  end
end

File: Práctica5/java/src/deepspace/BetaPowerEfficientSpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

/**
 * Representa una estación espacial eficiente con mejor tecnológia 
 * de eficiencia, llamada beta.
 * La estación eficiencia beta recibe mayor potencia de disparo
 */
public class BetaPowerEfficientSpaceStation extends PowerEfficientSpaceStation {
    /**
     * Eficiencia extra de la estación
     */
    private final float EXTRAEFFICIENCY = 1.2f;
    
    /**
     * Random
     */
    private Dice dice;

    /**
     * Constructor
     * @param station estación espacial a convertir en eficiente con tecnología beta
     */
    public BetaPowerEfficientSpaceStation(SpaceStation station) {
        super(station);
        dice = new Dice();
    }
    
    /**
     * Disparo con probabilidad de usar tecnología beta
     * @return potencia de disparo
     */
    @Override
    public float fire(){
        if (dice.extraEfficiency())
            return super.fire()*EXTRAEFFICIENCY;
        else
            return super.fire();
    }
    
    // -------------------------------------------------------------------------
    // String representation, UI version
    // -------------------------------------------------------------------------
    
    /**
     * String representation of the object.
     * @return string representation
     */
    @Override
    public String toString() {
        String message = "(Beta)" + super.toString();
        return message;
    }
    
    /**
     * To UI.
     * @return 
     */
    @Override
    public BetaPowerEfficientSpaceStationToUI getUIversion() {
        return new BetaPowerEfficientSpaceStationToUI(this);
    }
    
}
    
    